packages/next/src/build/index.ts

```diff
@@ -1518,6 +1631,11 @@ export default async function build(
         )
       }
 
+      // For pages directory, we run type checking after route collection but before build.
+      if (!appDir && !isCompileMode) {
+        await startTypeChecking(typeCheckingOptions)
```

Comment:
We could move this right after the routes type manifest is written to make it clear it should be eager but relies on that. 

Also let's keep the previous comment for context in addition to this.